### PR TITLE
Refactors initialization to use `setup()`

### PR DESCRIPTION
**Pull request by billbrod**
This PR adds a `setup` method to the OptimizedSynthesized classes: Metamer, MADCompetition, and Geodesic. setup handles the initialization of the synthesis image(s), optimizer, and scheduler. It does not need to be called by the user, in which case we call it the first time `synthesize` is called. All arguments are optional, in which case we do the behavior, which is unchanged from before and depends on the synthesis class. Splitting off setup like this makes the interaction with load easier: if load is called, setup must be called after it and can only be used to setup the optimizer/scheduler, if necessary. This also simplifies the logic of our optimizer and scheduler, and allows the user to pass an un-initialized version of the objects.

Also adds a warning when load is called if `setup()` will need to be called to initialize optimizer or scheduler.

This follows up from #313, because I realized that calling load with different values for initial_image or initial_noise (for Metamer / MADCompetition) wouldn't result in an error, which felt unlike the behavior of the rest of the initializing args, and unlike Geodesic, where a different value for initial_sequence would result in an error. 

Still to do:
- [x] Make docstrings more informative
- [x] Update docs

Example usages with metamer:

All code blocks start with this:
```python
img = po.data.einstein()
model = po.simul.Gaussian(30)
po.tools.remove_grad(model)
```

- Default case:
  ```python
  met = po.synth.Metamer(img, model)
  met.synthesize(10)
  ```
- Set initial image:
  ```python
  met = po.synth.Metamer(img, model)
  met.setup(po.data.curie())
  met.synthesize(10)
  ```
- Set optimizer (kwargs are optional):
  ```python
  met = po.synth.Metamer(img, model)
  met.setup(optimizer=torch.optim.SGD, optimizer_kwargs={"lr": 10})
  met.synthesize(10)
  ```
- Set scheduler (kwargs are optional):
  ```python
  met = po.synth.Metamer(img, model)
  met.setup(scheduler=torch.optim.lr_scheduler.ReduceLROnPlateau, 
            scheduler_kwargs={"factor": 1e-3})
  met.synthesize(10)
  ```
- with save/load (after load, when calling load, only optimizer and scheduler args can be set, initial_image or the kwargs will raise a ValueError; their values are set during load):
  ```python
  met = po.synth.Metamer(img, model)
  met.setup(po.data.curie(), optimizer=torch.optim.SGD, optimizer_kwargs={"lr": 10})
  met.synthesize(10)
  met.save("test.pt")
  met = po.synth.Metamer(img, model)
  met.load("test.pt")
  # this is necessary because we don't know how to initialize the optimizer object
  met.setup(optimizer=torch.optim.SGD)
  met.synthesize(10)
  ```

### CODE REVIEW

**Comment by billbrod** [comment]
Documentation built by flatiron-jenkins at http://docs.plenoptic.org/docs//pulls/325

File: src/plenoptic/synthesize/geodesic.py [hunk 1]
```diff
@@ -204,7 +230,9 @@ def synthesize(
             )
         print(f"\n Stop criterion for pixel_change_norm = {stop_criterion:.5e}")
 
-        self._initialize_optimizer(optimizer, "_geodesic", 0.001)
+        # if setup hasn't been called manually, call it now.
+        if self._geodesic is None or isinstance(self._optimizer, tuple):
+            self.setup()
```
Comment (BalzaniEdoardo) [inline src/plenoptic/synthesize/geodesic.py:275]: After this, you'll set loaded to `False`, is that what you want at the end of synthesis or do you consider the model as "loaded" once synthesis has been performed?

**Review by BalzaniEdoardo — APPROVED** [review]
This looks good, I have just a question about the flag _loaded. 

Unrelated:
Is there a reason why you are using the notebooks directly instead of myst-nb? It is hard to read what's changed in the notebooks, it would be nice to be able to use the notebook changes as an example to better understand the before and after of the change

File: src/plenoptic/synthesize/geodesic.py [hunk 2]
```diff
@@ -130,14 +125,20 @@ def __init__(
...
         self._step_energy_dims = None
-        # call this to cache some initial calculations
-        self.objective_function()
+        self._geodesic_representation = None
+        self._most_recent_step_energy = None
 
-    def _initialize(self, initial_sequence, start, stop, n_steps):
+    def setup(
+        self,
+        initial_sequence: Literal["straight", "bridge"] | None = None,
```
[the remark this answers is not in the document]
Comment (billbrod) [reply src/plenoptic/synthesize/geodesic.py:137]: Ruff doesn't like it. From rule [UP007](https://docs.astral.sh/ruff/rules/non-pep604-annotation-union/), which prefers it as more "concise and readable" (given that our minimum python version is 3.10).

**Comment by billbrod** [comment]
> Unrelated: Is there a reason why you are using the notebooks directly instead of myst-nb? It is hard to read what's changed in the notebooks, it would be nice to be able to use the notebook changes as an example to better understand the before and after of the change

I am switching over to myst-nb soon, it's my next major PR after this one.

File: src/plenoptic/synthesize/geodesic.py (hunk 1 above)
Comment (billbrod) [reply src/plenoptic/synthesize/geodesic.py:275]: the `_loaded` flag is only used to raise a more informative error message if someone calls `load()` and then `setup()` while trying to set `initial_image` / `initial_noise` / `initial_sequence`, e.g.,:

```
            if self._loaded:
                if initial_sequence is not None:
                    raise ValueError(
                        "Cannot set initial_sequence after calling load()!"
                    )
```

if they call `setup()` again after calling synthesize, then they'll get this error, which I think works:
```
                raise ValueError(
                    "setup() can only be called once and must be called"
                    " before synthesize()!"
                )
```

File: src/plenoptic/synthesize/geodesic.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (billbrod) [reply src/plenoptic/synthesize/geodesic.py:137]: I tend to agree, `Optional` (meaning a value of None is allowed) is confusable with "optional" (meaning there are default values), so typing it with `| None` is cleaner. See [typing docs](https://docs.python.org/3/library/typing.html#typing.Optional) which point this out.

[the remark this answers is not in the document]
Comment (BalzaniEdoardo) [reply src/plenoptic/synthesize/geodesic.py:137]: sounds good, I have a ton of Optional in nemos

[the remark this answers is not in the document]
Comment (billbrod) [reply src/plenoptic/synthesize/geodesic.py:137]: ruff can fix it all for you

File: src/plenoptic/synthesize/synthesis.py [hunk 3]
```diff
@@ -444,61 +464,50 @@ def _initialize_optimizer(
...
         """
-        synth_attr = getattr(self, synth_name)
-        if optimizer is None:
-            if self.optimizer is None:
-                self._optimizer = torch.optim.Adam(
-                    [synth_attr], lr=learning_rate, amsgrad=True
+        if isinstance(self.optimizer, tuple):
+            # then we're calling this after load()
+            if optimizer_kwargs is not None:
+                raise ValueError(
```
Comment (sjvenditto) [inline src/plenoptic/synthesize/synthesis.py:470]: Test for this error?

File: src/plenoptic/synthesize/synthesis.py [hunk 4]
```diff
@@ -526,8 +535,14 @@ def _initialize_scheduler(
         initialization). Then, we will pass the loss to scheduler.step when we call it.
 
         """
-        if scheduler is None:
-            if isinstance(self.scheduler, tuple):
+        if isinstance(self.scheduler, tuple):
+            # then we're calling this after load()
+            if scheduler_kwargs is not None:
+                raise ValueError(
```
Comment (sjvenditto) [inline src/plenoptic/synthesize/synthesis.py:541]: Test for this error?

**Comment by billbrod** [comment]
Added the tests you pointed out, plus modified my existing tests for the optimizer/scheduler to check that their class and parameters were set correctly.

**Review by sjvenditto — APPROVED** [review]
Looks good!
